lib/modal.fnl

```diff
@@ -0,0 +1,355 @@
+(local atom (require :lib.atom))
+(local statemachine (require :lib.statemachine))
+(local apps (require :lib.apps))
+(local {:call-when call-when
+        :concat    concat
+        :find      find
+        :filter    filter
+        :get       get
+        :has-some? has-some?
+        :join      join
+        :last      last
+        :map       map
+        :merge     merge
+        :noop      noop
+        :slice     slice
+        :tap       tap}
+       (require :lib.functional))
+(local {:align-columns align-columns}
+       (require :lib.text))
+(local {:action->fn action->fn
+        :bind-keys bind-keys}
+       (require :lib.bind))
+(local lifecycle (require :lib.lifecycle))
+
+(local log (hs.logger.new "\tmodal.fnl\t" "debug"))
+(var fsm nil)
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; General Utils
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(fn timeout
+  [f]
+  (let [task (hs.timer.doAfter 2 f)]
+    (fn destroy-task
+      []
+      (when task
+        (: task :stop)
+        nil))))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Event Dispatchers
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(fn activate-modal
+  [menu-key]
+  (fsm.dispatch :activate menu-key))
+
+
+(fn deactivate-modal
+  []
+  (fsm.dispatch :deactivate))
+
+
+(fn previous-modal
+  []
+  (fsm.dispatch :previous))
+
+
+(fn start-modal-timeout
+  []
+  (fsm.dispatch :start-timeout))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Set Key Bindings
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(fn create-action-trigger
+  [{:action action :repeatable repeatable :timeout timeout}]
+  (let [action-fn (action->fn action)]
+    (fn []
+      (if (and repeatable (~= timeout false))
+          (start-modal-timeout)
+          (not repeatable)
+          (deactivate-modal))
```

Comment:
this `if` statement looks confusing. Is there `do` block missing here, or something?

Reply:
That's fair. Unfortunately, fennel decided to make the `if` behave more like Clojure's `cond` where it handles pairs of conditions and lists to execute.

Three options come to mind:

1. Create an alias to `if` called `cond` to make it more clear when I'm intentionally using it in this way. It can be documented in `functional.fnl`.
2. Replace the single if with multiple `when` calls, which may be more appropriate to the behavior I had in mind.
    ```
    (when (and repeatable (~= timeout false)) 
      (start-modal-timeout))
    (when (not repeatable)
      (deactivate-modal))
     ```
3. Format it slightly differently like:
    ```
    (if (and repeatable  (~= timeout false)) (start-modal-timeout)
        (not repeatable)                     (deactivate-modal))
     ```